### PR TITLE
feat: add metadata, user caching, improve event dispatch

### LINKED ISSUE #216: Metadata System
### Confirmation

- [X] I have checked for similar issues.

### Problem

```java
void myMethod(User user) {
    MyUser myUser = userService.getUser(user.getId()); 
    Group group = groupService.getGroup(myUser.getId());
    // todo something with group
}
```
The current implementation of the framework requires multiple service calls to retrieve a user's group information, which can be inefficient and result in unnecessary database queries..

### Suggested solution

I suggest implementing a metadata system that allows developers to attach custom metadata to a user object, which can be retrieved in a single service call. This would allow for more efficient and streamlined access to user group information.

Code after the sugestion: 

```java
public static final MetadataKey<Integer> USER_ID = MetadataKey.of("user_id", Integer.class);

void myMethod(User user) {
    Group group = groupService.getGroup(user.getMetadata(USER_ID));
    // todo something with group
}
```

### Additional information

The suggested metadata system could be implemented using a key-value store, similar to the example provided in the code snippet.

### PR DESCRIPTION
**Pull request by joshuasing**
**Summary**
Add user metadata system, add user caching, improve event dispatch, improve audience provider.

Fixes #216

**Changes**
- Add user metadata system (fix #216)
- Add user caching
- Improve event dispatch
- Improve audience provider

**Checklist**
- [x] I acknowledge and agree to the terms of the [Developer Certificate of Origin](https://developercertificate.org/).
- [x] All contributed code can be distributed under the terms of the [MIT License](https://github.com/ChameleonFramework/Chameleon/blob/main/LICENSE).
- [x] I have read the [contributing guidelines](https://github.com/ChameleonFramework/Chameleon/blob/main/CONTRIBUTING.md) and [code of conduct](https://github.com/ChameleonFramework/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have checked the ["Allow edit from maintainers"](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) option.
- [x] I have added appropriate unit tests for my changes.

### CODE REVIEW

File: platform-nukkit/src/main/java/dev/hypera/chameleon/platform/nukkit/event/NukkitEventDispatcher.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+import dev.hypera.chameleon.user.User;
+import java.util.function.Consumer;
+import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
+import org.jetbrains.annotations.ApiStatus.Internal;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Nukkit event dispatcher.
+ */
+public final class NukkitEventDispatcher extends PlatformEventDispatcher implements Listener {
```
Comment (LooFifteen) [inline platform-nukkit/src/main/java/dev/hypera/chameleon/platform/nukkit/event/NukkitEventDispatcher.java:54]: Does this class unregister its listeners?

File: platform-sponge/src/main/java/dev/hypera/chameleon/platform/sponge/user/SpongeUserManager.java [hunk 2]
```diff
@@ -24,40 +24,48 @@
...
+                .plugin(chameleon.getPlatformPlugin().getPluginContainer())
+                .listener(event -> addUser(event.player().uniqueId(), event.player()))
+                .order(Order.EARLY).build()
+        );
+        Sponge.eventManager().registerListener(
+            EventListenerRegistration.builder(ServerSideConnectionEvent.Disconnect.class)
+                .plugin(chameleon.getPlatformPlugin().getPluginContainer())
+                .listener(event -> removeUser(event.player().uniqueId()))
+                .order(Order.LATE).build()
+        );
```
Comment (LooFifteen) [inline platform-sponge/src/main/java/dev/hypera/chameleon/platform/sponge/user/SpongeUserManager.java:68]: Are these unregistered either?

**Review by LooFifteen — CHANGES_REQUESTED** [review]
Looks like a load of listeners don't ever get unregistered, which could cause problems when people are using plugins like Plugman. We should be handling support for these kind of plugins because the end user of Chameleon does not have the option to.

File: platform-nukkit/src/main/java/dev/hypera/chameleon/platform/nukkit/event/NukkitEventDispatcher.java (hunk 1 above)
Comment (joshuasing) [reply platform-nukkit/src/main/java/dev/hypera/chameleon/platform/nukkit/event/NukkitEventDispatcher.java:54]: Nukkit does not appear to provide a way to unregister listeners:
https://javadoc.io/doc/cn.powernukkitx/powernukkitx/latest/cn/nukkit/plugin/PluginManager.html
(PowerNukkitX javadoc, I cannot find NukkitX javadoc)

Comment (joshuasing) [reply platform-nukkit/src/main/java/dev/hypera/chameleon/platform/nukkit/event/NukkitEventDispatcher.java:54]: Never mind, they have the exact same API as Bukkit: https://javadoc.io/doc/cn.powernukkitx/powernukkitx/latest/cn/nukkit/event/HandlerList.html

**Pull request merged**